python/pyspark/sql/tests.py

```diff
@@ -4052,6 +4066,323 @@ def test_unsupported_types(self):
+
+    def test_complex(self):
+        from pyspark.sql.functions import col, sum
+
+        df = self.data
+        plus_one = self.plus_one
+        plus_two = self.plus_two
+        sum_udf = self.sum_udf
+
+        result1 = (df.withColumn('v1', plus_one(df.v))
+                   .withColumn('v2', df.v + 2)
+                   .groupby('id')
+                   .agg(sum_udf(col('v')),
+                        sum_udf(col('v1') + 3),
+                        sum_udf(col('v2')) + 5,
+                        plus_one(sum_udf(col('v1'))),
+                        sum_udf(plus_one(col('v2'))))
+                   .sort('id')
+                   .toPandas())
+
+        expected1 = (df.withColumn('v1', df.v + 1)
+                     .withColumn('v2', df.v + 2)
+                     .groupby('id')
+                     .agg(sum(col('v')).alias('sum_udf(v)'),
+                          sum(col('v1') + 3).alias('sum_udf((v1 + 3))'),
+                          (sum(col('v2')) + 5).alias('(sum_udf(v2) + 5)'),
+                          plus_one(sum(col('v1'))).alias('plus_one(sum_udf(v1))'),
+                          sum(col('v2') + 1).alias('sum_udf(plus_one(v2))'))
+                     .sort('id')
+                     .toPandas())
+
+        result2 = (df.withColumn('v1', plus_one(df.v))
+                   .withColumn('v2', df.v + 2)
+                   .groupby('id')
+                   .agg(sum_udf(col('v')),
+                        sum_udf(col('v1') + 3),
+                        sum_udf(col('v2')) + 5,
+                        plus_two(sum_udf(col('v1'))),
+                        sum_udf(plus_two(col('v2'))))
+                   .sort('id')
+                   .toPandas())
+
+        expected2 = (df.withColumn('v1', df.v + 1)
+                     .withColumn('v2', df.v + 2)
+                     .groupby('id')
+                     .agg(sum(col('v')).alias('sum_udf(v)'),
+                          sum(col('v1') + 3).alias('sum_udf((v1 + 3))'),
+                          (sum(col('v2')) + 5).alias('(sum_udf(v2) + 5)'),
+                          plus_two(sum(col('v1'))).alias('plus_two(sum_udf(v1))'),
+                          sum(col('v2') + 2).alias('sum_udf(plus_two(v2))'))
+                     .sort('id')
+                     .toPandas())
+
+        result3 = (df.groupby('id')
+                   .agg(sum_udf(df.v).alias('v'))
+                   .groupby('id')
+                   .agg(sum_udf(col('v')).alias('sum_v'))
+                   .sort('id')
+                   .toPandas())
+
+        expected3 = (df.groupby('id')
+                     .agg(sum(df.v).alias('v'))
+                     .groupby('id')
+                     .agg(sum(col('v')).alias('sum_v'))
+                     .sort('id')
+                     .toPandas())
+
+        self.assertPandasEqual(expected1, result1)
+        self.assertPandasEqual(expected2, result2)
+        self.assertPandasEqual(expected3, result3)
+
```

Comment:
Can you add tests for complex/empty groupby like `GroupbyApplyTests.test_complex_groupby` or `GroupbyApplyTests.test_empty_groupby`?

Reply:
Added `test_complex_grouping`